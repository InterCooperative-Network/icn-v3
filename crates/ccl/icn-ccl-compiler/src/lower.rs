use icn_ccl_dsl::{DslModule, Proposal, Rule as DslRule, Value as DslValue};
use icn_ccl_parser::{CclParser, Rule};
use pest::iterators::{Pair, Pairs};
use pest::Parser;
use thiserror::Error;
use uuid::Uuid;

// Constant UUID for deterministic test snapshots
#[cfg(test)]
const TEST_UUID_STR: &str = "f0f1f2f3-f4f5-f6f7-f8f9-fafbfcfdfeff"; // Different from DSL test UUID

#[derive(Debug, Error)]
pub enum LowerError {
    #[error("parse error: {0}")]
    Parse(#[from] pest::error::Error<Rule>),
    #[error("unhandled rule: {0:?}")]
    Unhandled(Pair<'static, Rule>),
}

/// Primary entry‐point used by CLI & tests.
pub fn lower_str(src: &str) -> Result<Vec<DslModule>, LowerError> {
    let mut pairs = CclParser::parse(Rule::ccl, src)?;
    let ccl_root_pair = pairs.next().ok_or_else(|| {
        // This case should ideally not happen if parsing Rule::ccl was successful
        // and the grammar expects at least SOI/EOI or some content.
        // Creating a generic parse error if it does.
        pest::error::Error::new_from_span(
            pest::error::ErrorVariant::CustomError {
                message: "Expected a CCL root pair but found none.".to_string(),
            },
            pest::Span::new(src, 0, 0).unwrap(), // Dummy span
        )
    })?;
    Lowerer::default().lower(ccl_root_pair.into_inner())
}

#[derive(Default)]
struct Lowerer;

impl Lowerer {
    fn lower(&self, pairs: Pairs<'_, Rule>) -> Result<Vec<DslModule>, LowerError> {
        let mut modules = Vec::new();

        for pair in pairs {
            match pair.as_rule() {
                Rule::statement => {
                    // statement -> inner definition(s)
                    for inner in pair.into_inner() {
                        // Pass a mutable reference to modules
                        self.dispatch_def(&mut modules, inner)?;
                    }
                }
                // Handle cases where a definition might not be wrapped in a statement,
                // or for rules like EOI that are direct children of ccl_root_pair.into_inner()
                _ => self.dispatch_def(&mut modules, pair)?,
            }
        }
        Ok(modules)
    }

    fn dispatch_def(
        &self,
        modules: &mut Vec<DslModule>,
        pair: Pair<'_, Rule>,
    ) -> Result<(), LowerError> {
        match pair.as_rule() {
            Rule::proposal_def => {
                modules.push(DslModule::Proposal(self.lower_proposal(pair)?));
            }
            Rule::election_def => {
                modules.push(DslModule::Proposal(self.lower_election(pair)?));
            }
            // Top-level definitions from election.ccl and other templates - no-op for now
            Rule::roles_def |
            Rule::process_def |
            Rule::vacancies_def |
            Rule::actions_def |
            Rule::organization_def |
            Rule::governance_def |
            Rule::membership_def |
            Rule::budget_def |
            Rule::allocations_def |
            Rule::spending_rules_def |
            Rule::reporting_def => {
                // TODO: Implement lowering for these rules
                // For now, consuming the pair and doing nothing allows tests to proceed
            }

            Rule::EOI => {} // EOI will be the last item from ccl_root_pair.into_inner()
            _other => {
                // Avoid transmuting a Pair that might be 'static if it came from the top level
                // For now, let's create a new owned Pair for the error if it's not proposal_def or EOI/SOI
                // A better approach would be to ensure all handled rules are exhaustive or make Unhandled take Pair<'_, Rule>
                // but that requires changing LowerError and its usage.
                // For simplicity in this step, we create an owned string representation for the error.
                // This is a simplification; a robust solution would handle lifetimes carefully.
                return Err(LowerError::Unhandled(unsafe { std::mem::transmute(pair) }));
            }
        }
        Ok(())
    }

    fn lower_proposal(&self, pair: Pair<'_, Rule>) -> Result<Proposal, LowerError> {
        let pair_span = pair.as_span();
        let mut proposal_specific_pairs = pair.into_inner(); // These are specific to proposal_def

        let title = proposal_specific_pairs
            .next()
            .ok_or_else(|| LowerError::Parse(pest::error::Error::new_from_span(
                pest::error::ErrorVariant::CustomError { message: "Proposal missing title".to_string() },
                pair_span,
            )))?
            .as_str()
            .trim_matches('"')
            .to_owned();

        let block_pair = proposal_specific_pairs.next().ok_or_else(|| LowerError::Parse(pest::error::Error::new_from_span(
            pest::error::ErrorVariant::CustomError { message: "Proposal missing block".to_string() },
            pair_span,
        )))?;

        let mut description_body = String::new();
        let mut dsl_rules = Vec::<DslRule>::new();

        if block_pair.as_rule() == Rule::block {
            for statement_pair in block_pair.into_inner() { // These are statements inside the block
                if statement_pair.as_rule() == Rule::statement {
                    // A statement should have one inner actual definition (e.g., any_statement)
                    if let Some(inner_def_pair) = statement_pair.into_inner().next() {
                        match inner_def_pair.as_rule() {
                            Rule::any_statement => {
                                let mut field_parts = inner_def_pair.into_inner();
                                let key_pair = field_parts.next();
                                let value_pair_outer = field_parts.next();

                                if let (Some(key), Some(val_outer)) = (key_pair, value_pair_outer) {
                                    let key_str = key.as_str();
                                    // val_outer is a Rule::value, need to get its actual inner string/number etc.
                                    if let Some(val_inner) = val_outer.into_inner().next() {
                                        let value_str = val_inner.as_str().trim_matches('"').to_string();
                                        if key_str == "description" {
                                            description_body = value_str;
                                        } else if key_str == "version" {
                                            dsl_rules.push(DslRule {
                                                key: key_str.to_string(),
                                                value: DslValue::String(value_str),
                                            });
                                        } else {
                                            // Other fields can become generic rules for now
                                            dsl_rules.push(DslRule {
                                                key: key_str.to_string(),
                                                value: DslValue::String(value_str), // Or more specific DslValue type
                                            });
                                        }
                                    }
                                }
                            }
                            _ => { /* Unhandled statement type in block */ }
                        }
                    }
                }
            }
        }
        Ok(self.build_stub_proposal(title, description_body, dsl_rules))
    }

    fn lower_election(&self, pair: Pair<'_, Rule>) -> Result<Proposal, LowerError> {
        let pair_span = pair.as_span();
        let mut election_specific_pairs = pair.into_inner(); // These are specific to election_def

        let title = election_specific_pairs
            .next()
            .ok_or_else(|| LowerError::Parse(pest::error::Error::new_from_span(
                pest::error::ErrorVariant::CustomError { message: "Election missing title".to_string() },
                pair_span,
            )))?
            .as_str()
            .trim_matches('"')
            .to_owned();

        // The next pair should be the block
        let block_pair = election_specific_pairs.next().ok_or_else(|| LowerError::Parse(pest::error::Error::new_from_span(
            pest::error::ErrorVariant::CustomError { message: "Election missing block".to_string() },
            pair_span, // Use the original election_def span for error reporting
        )))?;

        let mut description_body = String::new();
        let mut dsl_rules = Vec::<DslRule>::new();

        if block_pair.as_rule() == Rule::block {
            for statement_pair in block_pair.into_inner() { // These are statements inside the block
                if statement_pair.as_rule() == Rule::statement {
                    // A statement should have one inner actual definition (e.g., any_statement)
                    if let Some(inner_def_pair) = statement_pair.into_inner().next() {
                        match inner_def_pair.as_rule() {
                            Rule::any_statement => {
                                let mut field_parts = inner_def_pair.into_inner();
                                let key_pair = field_parts.next();
                                let value_pair_outer = field_parts.next();

                                if let (Some(key), Some(val_outer)) = (key_pair, value_pair_outer) {
                                    let key_str = key.as_str();
                                    // val_outer is a Rule::value, need to get its actual inner string/number etc.
                                    if let Some(val_inner) = val_outer.into_inner().next() {
                                        let value_str = val_inner.as_str().trim_matches('"').to_string();
                                        if key_str == "description" {
                                            description_body = value_str;
                                        } else if key_str == "version" {
                                            dsl_rules.push(DslRule {
                                                key: key_str.to_string(),
                                                value: DslValue::String(value_str),
                                            });
                                        } else {
                                            // Other fields can become generic rules for now
                                            dsl_rules.push(DslRule {
                                                key: key_str.to_string(),
                                                value: DslValue::String(value_str), // Or more specific DslValue type
                                            });
                                        }
                                    }
                                }
                            }
                            _ => { /* Unhandled statement type in block */ }
                        }
                    }
                }
            }
        }
        // Use the extracted title, description (as body), and version (as a rule)
        Ok(self.build_stub_proposal(title, description_body, dsl_rules))
    }

    fn build_stub_proposal(&self, title: String, body: String, rules: Vec<DslRule>) -> Proposal {
        let id = {
            #[cfg(test)]
            { Uuid::parse_str(TEST_UUID_STR).unwrap() } 
            #[cfg(not(test))]
            { Uuid::new_v4() } 
        };

        Proposal {
            id,
            title,
            body,
            author: "unknown".into(),
            created_at: 0,
            rules, // Use passed in rules
        }
    }
} 