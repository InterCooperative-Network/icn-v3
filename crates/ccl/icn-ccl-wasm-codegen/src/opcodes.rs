//'!' Extremely-rough first cut at a host-call opcode list.
//'!' Everything will get revisited once we know the real WASM ABI.

use serde::{Deserialize, Serialize};

// Opcode represents a single operation in a compiled ICN program.
// These are high-level opcodes, not raw WASM instructions.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub enum Opcode {
    // high-level
    CreateProposal { title: String, version: Option<String> },
    OnEvent { event: String },

    // actions
    MintToken { res_type: String, amount: u64, recipient: Option<String> },
    AnchorData { path: Option<String>, data_ref: String },
    CallHost { fn_name: String, args: Vec<String> },

    // control flow
    If { condition: String },
    Else,
    EndIf,

    // misc
    RangeCheck { start: f64, end: f64 },
    BeginSection { kind: String, title: Option<String> },
    EndSection,

    /// Simple key/value pair that doesn't warrant its own opcode.
    /// `value_json` is always valid JSON (even for strings â€“ we quote them).
    SetProperty { key: String, value_json: String },
    Todo(String),
}

// Program is a sequence of opcodes, the result of compiling a DslModule list.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Program {
    pub ops: Vec<Opcode>,
}

impl Program {
    pub fn new(ops: Vec<Opcode>) -> Self {
        Program { ops }
    }
} 