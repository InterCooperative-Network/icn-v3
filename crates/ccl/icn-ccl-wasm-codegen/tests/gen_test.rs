use icn_ccl_compiler::lower::lower_str;
use icn_ccl_wasm_codegen::WasmGenerator;
use icn_ccl_wasm_codegen::opcodes::Program;
use insta::assert_json_snapshot;

// Helper to load CCL, parse, lower, and generate opcodes
fn modules_from_ccl_string(ccl_string: &str) -> Program {
    let modules = lower_str(ccl_string).unwrap_or_else(|e| {
        panic!("Failed to lower CCL string to DSL: {:#?}", e);
    });
    let generator = WasmGenerator::new();
    generator.generate(modules) // Call generate on the instance
}

macro_rules! snapshot_file {
    ($name:ident, $path:expr) => {
        #[test]
        fn $name() {
            let src = include_str!($path);
            let program_ops = modules_from_ccl_string(src);
            assert_json_snapshot!(stringify!($name), program_ops);
        }
    };
}

// Test cases
snapshot_file!(election_ops, "../../icn-ccl-parser/templates/election.ccl");
snapshot_file!(budget_ops, "../../icn-ccl-parser/templates/budget.ccl");
snapshot_file!(bylaws_ops, "../../icn-ccl-parser/templates/bylaws.ccl"); 