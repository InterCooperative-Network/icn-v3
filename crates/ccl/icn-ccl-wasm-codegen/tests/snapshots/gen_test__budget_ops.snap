---
source: crates/ccl/icn-ccl-wasm-codegen/tests/gen_test.rs
expression: program_ops
---
{
  "ops": [
    {
      "CreateProposal": {
        "title": "cooperative_budget",
        "version": "0.0.0-unknown"
      }
    },
    {
      "SetProperty": {
        "key": "version",
        "value_json": "\"1.0.0\""
      }
    },
    {
      "SetProperty": {
        "key": "currency",
        "value_json": "\"USD\""
      }
    },
    {
      "SetProperty": {
        "key": "period",
        "value_json": "\"annual\""
      }
    },
    {
      "BeginSection": {
        "kind": "allocations",
        "title": null
      }
    },
    {
      "SetProperty": {
        "key": "category",
        "value_json": "\"operations\""
      }
    },
    {
      "SetProperty": {
        "key": "category",
        "value_json": "\"projects\""
      }
    },
    {
      "SetProperty": {
        "key": "category",
        "value_json": "\"reserve\""
      }
    },
    {
      "SetProperty": {
        "key": "category",
        "value_json": "\"community\""
      }
    },
    "EndSection",
    {
      "BeginSection": {
        "kind": "spending_rules",
        "title": null
      }
    },
    {
      "BeginSection": {
        "kind": "range_0_500",
        "title": "range_0_500"
      }
    },
    {
      "SetProperty": {
        "key": "approvers",
        "value_json": "1.0"
      }
    },
    {
      "SetProperty": {
        "key": "required_role",
        "value_json": "\"steward\""
      }
    },
    "EndSection",
    {
      "BeginSection": {
        "kind": "range_501_5000",
        "title": "range_501_5000"
      }
    },
    {
      "SetProperty": {
        "key": "approvers",
        "value_json": "2.0"
      }
    },
    {
      "SetProperty": {
        "key": "required_role",
        "value_json": "\"finance_committee\""
      }
    },
    "EndSection",
    {
      "BeginSection": {
        "kind": "range_5001_999999999",
        "title": "range_5001_999999999"
      }
    },
    {
      "SetProperty": {
        "key": "approval_threshold",
        "value_json": "0.6"
      }
    },
    {
      "SetProperty": {
        "key": "quorum",
        "value_json": "0.5"
      }
    },
    "EndSection",
    "EndSection",
    {
      "BeginSection": {
        "kind": "reporting",
        "title": null
      }
    },
    {
      "SetProperty": {
        "key": "frequency",
        "value_json": "\"monthly\""
      }
    },
    {
      "SetProperty": {
        "key": "fields",
        "value_json": "[\"category\",\"subcategory\",\"amount\",\"remaining_budget\",\"percentage_used\"]"
      }
    },
    {
      "SetProperty": {
        "key": "fields",
        "value_json": "[\"date\",\"amount\",\"category\",\"description\",\"authorized_by\"]"
      }
    },
    "EndSection",
    {
      "OnEvent": {
        "event": "expense.approved"
      }
    },
    {
      "MintToken": {
        "res_type": "expense_receipt",
        "amount": 1,
        "recipient": "ctx.submitter_id"
      }
    },
    {
      "AnchorData": {
        "path": "finance/expenses",
        "data_ref": "map_content_placeholder_[Rule { key: \"id\", value: String(\"ctx.expense_id\") }, Rule { key: \"details\", value: String(\"ctx.expense_details\") }, Rule { key: \"status\", value: String(\"approved\") }, Rule { key: \"approved_at\", value: Map([Rule { key: \"function_name\", value: String(\"timestamp\") }, Rule { key: \"args\", value: Map([]) }]) }]"
      }
    },
    {
      "OnEvent": {
        "event": "budget.proposed"
      }
    },
    {
      "MintToken": {
        "res_type": "budget_proposal",
        "amount": 1,
        "recipient": "ctx.proposer_id"
      }
    }
  ]
}
